README.md
spell_checker
==========

A simple spell checker implementation in Dart. For now it only finds the single words from a given 
dictionary. Algorithm is different than Peter Norvig's implementation (http://norvig.com/spell-correct.html). This implementation 
is more complicated but probably much faster (finds several thousands matches in a second). 
It uses dynamic decoding over a simple trie generated from the dictionary. System finds words with a distance to the input.
Deletions, insertions, substitutions and transpositions are supported.  

	import 'package:dart_spell/dart_spell.dart';
	
	...
	// optional distance parameter. Default is 1.0
	var checker = new SingleWordSpellChecker(distance:1.0);
	
	var dictionary = ["apple", "apples", "pear", "ear"];
	checker.addWords(dictionary);	
	
	List<Result> matches = checker.find("apple");
	print(matches);

	Output:
	[apple:0.0, apples:1.0]
	
##TODO
* Add less substitution penalty for near keys in keyboard layout.
* Add language model support so that it gives more logical suggestions.
* Add multi word spell suggestion with space and out of vocabulary word handling.  


CHANGELOG.md
## 0.1.2+3

Change Repo

## 0.1.2+2

Add example, comments, format to improve health

## 0.1.2+1

Change map to splaymap to avoid final sort

## 0.1.2

Remove comparable from Result

## 0.1.1+8

Update to effective dart and format with dartfmt

## 0.1.1+7

Make addChar private

## 0.1.1+6

Update addChar function header

## 0.1.1+5

Make root private to Search, make addChar iterative instead of recursive

## 0.1.1+4

Swap out List for Iterable in addWords

## 0.1.1+3

Remove uses of new to match modern dart

## 0.1.1+2

Remove unncessary hashing/index code

## 0.1.1+1

Dart 2.0 clean-up

## 0.1.1

Dart 1.0 clean-up.

## 0.1.0

Initial release


test/single_word_spell_checker_test.dart
library single_word_spell_checker_test;

import 'dart:math';
import 'package:test/test.dart';
import 'package:spell_checker/spell_checker.dart';
import 'package:list_english_words/list_english_words.dart';

final Random r = Random(0xbeef);

void main() {
  group('Test Word List', () {
    final checker = SingleWordSpellChecker(distance: 1.0);
    checker.addWords(list_english_words);

    test('All Words In List Come Back First', () {
      for(final s in list_english_words) {
        expect(checker.find(s)[0].word == s, isTrue);
      }
    });
  });

  group('Test Short List', () {
    final checker = SingleWordSpellChecker(distance: 1.0);
    checker.addWords(['apple', 'apply', 'applesauce', 'applause', 'pear']);
    final str = 'apple';
    test('All Variations', () {
      final delete = randomDelete(str, 1);
      for (final s in delete) {
        final findList = checker.find(s);
        final subSize = findList.length < 5 ? findList.length : 5;
        final testStrs =
            findList.sublist(0, subSize).map((e) => e.word).toList();
        expect(testStrs.contains(str), isTrue);
      }
      final insert = randomInsert(str, 1);
      for (var s in insert) {
        final findList = checker.find(s);
        final subSize = findList.length < 5 ? findList.length : 5;
        final testStrs =
            findList.sublist(0, subSize).map((e) => e.word).toList();
        expect(testStrs.contains(str), isTrue);
      }
      final substitute = randomSubstitute(str, 1);
      for (var s in substitute) {
        final findList = checker.find(s);
        final subSize = findList.length < 5 ? findList.length : 5;
        final testStrs =
            findList.sublist(0, subSize).map((e) => e.word).toList();
        expect(testStrs.contains(str), isTrue);
      }
      final transposition = transpositions(str);
      for (var s in transposition) {
        final findList = checker.find(s);
        final subSize = findList.length < 5 ? findList.length : 5;
        final testStrs =
            findList.sublist(0, subSize).map((e) => e.word).toList();
        expect(testStrs.contains(str), isTrue);
      }
    });
  });
}

Set<String> randomDelete(String input, int d) {
  final result = <String>{};
  for (var i = 0; i < 100; i++) {
    // ignore: omit_local_variable_types
    List<int> sb = List.from(input.codeUnits, growable: true);
    for (var j = 0; j < d; j++) {
      sb.removeAt(r.nextInt(sb.length));
    }
    result.add(String.fromCharCodes(sb));
  }
  return result;
}

Set<String> randomInsert(String input, int d) {
  final result = <String>{};
  for (var i = 0; i < 100; i++) {
    // ignore: omit_local_variable_types
    List<int> sb = List.from(input.codeUnits, growable: true);
    for (var j = 0; j < d; j++) {
      sb.insert(r.nextInt(sb.length + 1), 'x'.codeUnitAt(0));
    }
    result.add(String.fromCharCodes(sb));
  }
  return result;
}

Set<String> randomSubstitute(String input, int d) {
  final result = <String>{};
  for (var i = 0; i < 100; i++) {
    // ignore: omit_local_variable_types
    List<int> sb = List.from(input.codeUnits, growable: true);
    for (var j = 0; j < d; j++) {
      var start = r.nextInt(sb.length);
      sb[start] = 'x'.codeUnitAt(0);
    }
    result.add(String.fromCharCodes(sb));
  }
  return result;
}

Set<String> transpositions(String input) {
  final result = <String>{};
  for (var i = 0; i < input.length - 1; i++) {
    // ignore: omit_local_variable_types
    List<int> sb = List.from(input.codeUnits, growable: true);
    var tmp = sb[i];
    sb[i] = sb[i + 1];
    sb[i + 1] = tmp;
    result.add(String.fromCharCodes(sb));
  }
  return result;
}


lib/dart_spell.dart
library dart_spell;

const int CHARACTER_SPACE = 32;
const int CHARACTER_RANGE_LOW = 32;
const int CHARACTER_RANGE_HIGH = 126;
const int NUMBER_OF_CHARACTERS =
    CHARACTER_RANGE_HIGH - CHARACTER_RANGE_LOW - 26;

bool isNull(Object obj) {
  return obj == null;
}

bool isChar(int char) {
  return !isNull(char) &&
      char >= CHARACTER_RANGE_LOW &&
      char <= CHARACTER_RANGE_HIGH;
}

extension ContainsCodeUnit on String {
  bool containsCodeUnit(int i) => runes.contains(i);
}

///
/// Simple dictionary based spell checker.
///
class SingleWordSpellChecker {
  final bool checkNearKeySubstitution = false;

  final double INSERTION_PENALTY = 0.4;
  final double DELETION_PENALTY = 0.6;
  final double SUBSTITUTION_PENALTY = 0.6;
  final double TRANSPOSITION_PENALTY = 0.6;

  //TODO: not used yet.
  final double NEAR_KEY_SUBSTITUTION_PENALTY = 0.5;
  final Map<int, String> nearKeyMap = <int, String>{};
  Map<String, double> hypotheses;

  double distance;
  _Node _root;

  SingleWordSpellChecker({num distance = 1}) {
    if (distance != null) this.distance = distance;
    _root = _Node(0);
  }

  void addWord(String word) => _addChar(word, word);
  void addWords(Iterable<String> words) => words?.forEach(addWord);

  void _addChar(String word, String actual) {
    var tmpNode = _root;
    for (final rune in word?.toLowerCase()?.runes) {
      // TODO: Add tests for this
      // TODO: Make custom exception
      if (!isChar(rune)) {
        throw Exception(
            'Invalid character added to spell checker, code unit: $rune}');
      }
      tmpNode = tmpNode?.addChild(rune);
    }
    tmpNode.word = actual;
  }

  bool isWord(String input) {
    var tmpNode = _root;
    for(final rune in input.runes) {
      if(tmpNode.hasChild(rune)) {
        tmpNode = tmpNode.getChild(rune);
      } else {
        return false;
      }
    }
    return tmpNode.endsWord;
  }

  List<Result> find(String input) {
    if(isWord(input)) return [Result(input, 0)];
    final lowered = input.toLowerCase();
    hypotheses = <String, double>{};

    final hyp = _Hypothesis(_root, 0.0, -1);
    var next = _expand(hyp, lowered);
    while (next.isNotEmpty) {
      final expanded = next.map((hyp) => _expand(hyp, lowered));
      next = expanded.reduce((e, v) => v?.union(e));
    }

    return hypotheses.keys.map((key) => Result(key, hypotheses[key])).toList()
      ..sort();
  }

  Set<_Hypothesis> _noError(_Hypothesis hypothesis, String input) {
    final nextIndex = hypothesis.index + 1;
    final hypNode = hypothesis.node;
    final newHypotheses = <_Hypothesis>{};

    if (nextIndex < input.length) {
      if (hypNode.hasChild(input.codeUnitAt(nextIndex))) {
        final hyp = hypothesis.getNewMoveForward(
            hypNode.getChild(input.codeUnitAt(nextIndex)), 0.0);
        newHypotheses.add(hyp);
      }
    } else {
      if (hypothesis.isWord) {
        _addHypothesis(hypothesis);
      }
    }
    return newHypotheses;
  }

  Set<_Hypothesis> _handleNearKey(
      _Hypothesis hypothesis, String input, _Node childNode) {
    final nextIndex = hypothesis.index + 1;
    final newHypotheses = <_Hypothesis>{};

    final nextChar = input.codeUnitAt(nextIndex);
    if (childNode.chr != nextChar) {
      final nearCharactersString = nearKeyMap[childNode.chr];
      if (nearCharactersString != null &&
          nearCharactersString.containsCodeUnit(nextChar)) {
        //NEAR_KEY_SUBSTITUTION_PENALTY;
      }
    }

    return newHypotheses;
  }

  Set<_Hypothesis> _handleSubstitution(
      _Hypothesis hypothesis, String input, _Node childNode) {
    final nextIndex = hypothesis.index + 1;
    final hypDist = hypothesis.distance;
    final newHypotheses = <_Hypothesis>{};

    if (hypDist + SUBSTITUTION_PENALTY <= distance) {
      // TODO consider double add for noError
      final hyp = hypothesis.getNewMoveForward(childNode, SUBSTITUTION_PENALTY);
      if (nextIndex == input.length - 1) {
        if (hyp.isWord) {
          _addHypothesis(hyp);
        }
      } else {
        newHypotheses.add(hyp);
      }
    }

    return newHypotheses;
  }

  Set<_Hypothesis> _substitution(_Hypothesis hypothesis, String input) {
    final nextIndex = hypothesis.index + 1;
    final children = hypothesis.node.children;
    final newHypotheses = <_Hypothesis>{};

// substitution
    if (nextIndex < input.length) {
      for (final childNode in children) {
        if (checkNearKeySubstitution) {
          newHypotheses.addAll(_handleNearKey(hypothesis, input, childNode));
        } else {
          newHypotheses
              .addAll(_handleSubstitution(hypothesis, input, childNode));
        }
      }
    }
    return newHypotheses;
  }

  Set<_Hypothesis> _deletion(_Hypothesis hypothesis, String input) {
    final hypNode = hypothesis.node;
    final newHypotheses = <_Hypothesis>{};

    newHypotheses.add(hypothesis.getNewMoveForward(hypNode, DELETION_PENALTY));
    return newHypotheses;
  }

  Set<_Hypothesis> _insertion(_Hypothesis hypothesis, String input) {
    final hypNode = hypothesis.node;
    final newHypotheses = <_Hypothesis>{};

    // insertion
    for (final childNode in hypNode.children) {
      newHypotheses.add(
          hypothesis.getNew(childNode, INSERTION_PENALTY, hypothesis.index));
    }

    return newHypotheses;
  }

  Set<_Hypothesis> _transposition(_Hypothesis hypothesis, String input) {
    final hypNode = hypothesis.node;
    final nextIndex = hypothesis.index + 1;
    final newHypotheses = <_Hypothesis>{};

    if (nextIndex < input.length - 1) {
      final transpose = input.codeUnitAt(nextIndex + 1);
      final nextNode = hypNode.getChild(transpose);
      final nextChar = input.codeUnitAt(nextIndex);
      if (hypNode.hasChild(transpose) && nextNode.hasChild(nextChar)) {
        final hyp = hypothesis.getNew(
            nextNode.getChild(nextChar), TRANSPOSITION_PENALTY, nextIndex + 1);
        if (nextIndex == input.length - 1) {
          if (hyp.isWord) {
            _addHypothesis(hyp);
          }
        } else {
          newHypotheses.add(hyp);
        }
      }
    }
    return newHypotheses;
  }

  Set<_Hypothesis> _expand(_Hypothesis hypothesis, String input) {
    final newHypotheses = <_Hypothesis>{};
    final hypDist = hypothesis.distance;

    // no-error
    newHypotheses.addAll(_noError(hypothesis, input));

    // we don't need to explore further if we reached to max penalty
    if (hypDist >= distance) {
      return newHypotheses;
    } else {
      newHypotheses.addAll(_substitution(hypothesis, input));
    }

    if (hypDist + DELETION_PENALTY > distance) {
      return newHypotheses;
    } else {
      newHypotheses.addAll(_deletion(hypothesis, input));
    }

    newHypotheses.addAll(_insertion(hypothesis, input));

    // transposition
    newHypotheses.addAll(_transposition(hypothesis, input));

    return newHypotheses;
  }

  void _addHypothesis(_Hypothesis hypToAdd) {
    final hypWord = hypToAdd.node.word;
    if (isNull(hypWord)) return;
    if(hypotheses.containsKey(hypWord)) {
      if(hypToAdd.distance < hypotheses[hypWord]) {
        hypotheses.update(hypWord, (val) => hypToAdd.distance);
      }
    } else {
      hypotheses.putIfAbsent(hypWord, () => hypToAdd.distance);
    }
  }
}

class _Node {
  final int chr;
  final Map<int, _Node> nodes = <int, _Node>{};

  String _word;
  set word(String word) => _word = word ?? _word;
  String get word => _word;

  _Node(this.chr);

  Iterable<_Node> get children => nodes.values;

  bool hasChild(int c) => nodes.containsKey(c);
  _Node getChild(int c) => nodes[c];
  _Node addChild(int c) => nodes.putIfAbsent(c, () => _Node(c));
  bool get endsWord => !isNull(_word);

  @override
  int get hashCode {
    return chr;
  }
}

// Hypothesis: That the node held in this string ends a word that could be a correct spelling
class _Hypothesis {
  final _Node node;
  final double distance;
  final int index;

  const _Hypothesis(this.node, this.distance, this.index);

  const _Hypothesis.initialNode(this.node, this.distance) : index = -1;

  //  _Hypothesis
  _Hypothesis getNewMoveForward(_Node node, double penaltyToAdd) {
    return _Hypothesis(node, distance + penaltyToAdd, index + 1);
  }

  _Hypothesis getNew(_Node node, double penaltyToAdd, int index) {
    return _Hypothesis(node, distance + penaltyToAdd, index);
  }

  bool get isWord => node.endsWord;

  @override
  int get hashCode {
    var result = node.hashCode;
    result= result*31+distance.hashCode;
    result= result*31+index;
    return result;
  }

  @override
  bool operator ==(other) {
    if (other is! _Hypothesis) return false;
    return index == other.index &&
        distance.compareTo(other.distance) == 0 &&
        node == other.node;
  }
}

class Result implements Comparable<Result> {
  final String word;
  final double distance;

  const Result(this.word, this.distance);

  @override
  int compareTo(Result other) => distance.compareTo(other.distance);

  @override
  String toString() => '$word:$distance';
}


lib/spell_checker.dart
library dart_spell;

import 'dart:collection';

import 'package:spell_checker/utilities.dart';

///
/// Simple dictionary based spell checker.
///
class SingleWordSpellChecker {
  final bool checkNearKeySubstitution = false;

  final double INSERTION_PENALTY = 0.4;
  final double DELETION_PENALTY = 0.6;
  final double SUBSTITUTION_PENALTY = 0.6;
  final double TRANSPOSITION_PENALTY = 0.6;

  //TODO: not used yet.
  final double NEAR_KEY_SUBSTITUTION_PENALTY = 0.5;
  final Map<int, String> nearKeyMap = <int, String>{};
  SplayTreeMap<String, double> hypotheses;

  double distance;
  _Node _root;

  SingleWordSpellChecker({num distance = 1}) {
    if (distance != null) this.distance = distance;
    _root = _Node(0);
  }

  void addWord(String word) => _addChar(word, word);

  void addWords(Iterable<String> words) => words?.forEach(addWord);

  void _addChar(String word, String actual) {
    var tmpNode = _root;
    for (final rune in word?.toLowerCase()?.runes) {
      // TODO: Add tests for this
      // TODO: Make custom exception
      if (!isChar(rune)) {
        throw Exception(
            'Invalid character added to spell checker, code unit: $rune}');
      }
      tmpNode = tmpNode?.addChild(rune);
    }
    tmpNode.word = actual;
  }

  bool isWord(String input) {
    var tmpNode = _root;
    for (final rune in input.runes) {
      if (tmpNode.hasChild(rune)) {
        tmpNode = tmpNode.getChild(rune);
      } else {
        return false;
      }
    }
    return tmpNode.endsWord;
  }

  List<Result> find(String input) {
    if (isWord(input)) return [Result(input, 0)];
    final lowered = input.toLowerCase();
    hypotheses = SplayTreeMap<String, double>();

    final hyp = _Hypothesis(_root, 0.0, -1);
    var next = _expand(hyp, lowered);
    while (next.isNotEmpty) {
      final expanded = next.map((hyp) => _expand(hyp, lowered));
      next = expanded.reduce((e, v) => v?.union(e));
    }

    return hypotheses.keys.map((key) => Result(key, hypotheses[key])).toList();
  }

  Set<_Hypothesis> _noError(_Hypothesis hypothesis, String input) {
    final nextIndex = hypothesis.index + 1;
    final hypNode = hypothesis.node;
    final newHypotheses = <_Hypothesis>{};

    if (nextIndex < input.length) {
      if (hypNode.hasChild(input.codeUnitAt(nextIndex))) {
        final hyp = hypothesis.getNewMoveForward(
            hypNode.getChild(input.codeUnitAt(nextIndex)), 0.0);
        newHypotheses.add(hyp);
      }
    } else {
      if (hypothesis.isWord) {
        _addHypothesis(hypothesis);
      }
    }
    return newHypotheses;
  }

  Set<_Hypothesis> _handleNearKey(
      _Hypothesis hypothesis, String input, _Node childNode) {
    final nextIndex = hypothesis.index + 1;
    final newHypotheses = <_Hypothesis>{};

    final nextChar = input.codeUnitAt(nextIndex);
    if (childNode.chr != nextChar) {
      final nearCharactersString = nearKeyMap[childNode.chr];
      if (nearCharactersString != null &&
          nearCharactersString.containsCodeUnit(nextChar)) {
        //NEAR_KEY_SUBSTITUTION_PENALTY;
      }
    }

    return newHypotheses;
  }

  Set<_Hypothesis> _handleSubstitution(
      _Hypothesis hypothesis, String input, _Node childNode) {
    final nextIndex = hypothesis.index + 1;
    final hypDist = hypothesis.distance;
    final newHypotheses = <_Hypothesis>{};

    if (hypDist + SUBSTITUTION_PENALTY <= distance) {
      // TODO consider double add for noError
      final hyp = hypothesis.getNewMoveForward(childNode, SUBSTITUTION_PENALTY);
      if (nextIndex == input.length - 1) {
        if (hyp.isWord) {
          _addHypothesis(hyp);
        }
      } else {
        newHypotheses.add(hyp);
      }
    }

    return newHypotheses;
  }

  Set<_Hypothesis> _substitution(_Hypothesis hypothesis, String input) {
    final nextIndex = hypothesis.index + 1;
    final children = hypothesis.node.children;
    final newHypotheses = <_Hypothesis>{};

    // substitution
    if (nextIndex < input.length) {
      for (final childNode in children) {
        if (checkNearKeySubstitution) {
          newHypotheses.addAll(_handleNearKey(hypothesis, input, childNode));
        } else {
          newHypotheses
              .addAll(_handleSubstitution(hypothesis, input, childNode));
        }
      }
    }
    return newHypotheses;
  }

  Set<_Hypothesis> _deletion(_Hypothesis hypothesis, String input) {
    final hypNode = hypothesis.node;
    final newHypotheses = <_Hypothesis>{};

    newHypotheses.add(hypothesis.getNewMoveForward(hypNode, DELETION_PENALTY));
    return newHypotheses;
  }

  Set<_Hypothesis> _insertion(_Hypothesis hypothesis, String input) {
    final hypNode = hypothesis.node;
    final newHypotheses = <_Hypothesis>{};

    // insertion
    for (final childNode in hypNode.children) {
      newHypotheses.add(
          hypothesis.getNew(childNode, INSERTION_PENALTY, hypothesis.index));
    }

    return newHypotheses;
  }

  Set<_Hypothesis> _transposition(_Hypothesis hypothesis, String input) {
    final hypNode = hypothesis.node;
    final nextIndex = hypothesis.index + 1;
    final newHypotheses = <_Hypothesis>{};

    if (nextIndex < input.length - 1) {
      final transpose = input.codeUnitAt(nextIndex + 1);
      final nextNode = hypNode.getChild(transpose);
      final nextChar = input.codeUnitAt(nextIndex);
      if (hypNode.hasChild(transpose) && nextNode.hasChild(nextChar)) {
        final hyp = hypothesis.getNew(
            nextNode.getChild(nextChar), TRANSPOSITION_PENALTY, nextIndex + 1);
        if (nextIndex == input.length - 1) {
          if (hyp.isWord) {
            _addHypothesis(hyp);
          }
        } else {
          newHypotheses.add(hyp);
        }
      }
    }
    return newHypotheses;
  }

  Set<_Hypothesis> _expand(_Hypothesis hypothesis, String input) {
    final newHypotheses = <_Hypothesis>{};
    final hypDist = hypothesis.distance;

    // no-error
    newHypotheses.addAll(_noError(hypothesis, input));

    // we don't need to explore further if we reached to max penalty
    if (hypDist >= distance) {
      return newHypotheses;
    } else {
      newHypotheses.addAll(_substitution(hypothesis, input));
    }

    if (hypDist + DELETION_PENALTY > distance) {
      return newHypotheses;
    } else {
      newHypotheses.addAll(_deletion(hypothesis, input));
    }

    newHypotheses.addAll(_insertion(hypothesis, input));

    // transposition
    newHypotheses.addAll(_transposition(hypothesis, input));

    return newHypotheses;
  }

  void _addHypothesis(_Hypothesis hypToAdd) {
    final hypWord = hypToAdd.node.word;
    if (isNull(hypWord)) return;
    if (hypotheses.containsKey(hypWord)) {
      if (hypToAdd.distance < hypotheses[hypWord]) {
        hypotheses.update(hypWord, (val) => hypToAdd.distance);
      }
    } else {
      hypotheses.putIfAbsent(hypWord, () => hypToAdd.distance);
    }
  }
}

class _Node {
  final int chr;
  final Map<int, _Node> nodes = <int, _Node>{};

  String _word;

  set word(String word) => _word = word ?? _word;

  String get word => _word;

  _Node(this.chr);

  Iterable<_Node> get children => nodes.values;

  bool hasChild(int c) => nodes.containsKey(c);

  _Node getChild(int c) => nodes[c];

  _Node addChild(int c) => nodes.putIfAbsent(c, () => _Node(c));

  bool get endsWord => !isNull(_word);

  @override
  int get hashCode {
    return chr;
  }
}

/// Hypothesis: That the node held in this string ends a word that could be a correct spelling
class _Hypothesis {
  final _Node node;
  final double distance;
  final int index;

  const _Hypothesis(this.node, this.distance, this.index);

  const _Hypothesis.initialNode(this.node, this.distance) : index = -1;

  //  _Hypothesis
  _Hypothesis getNewMoveForward(_Node node, double penaltyToAdd) {
    return _Hypothesis(node, distance + penaltyToAdd, index + 1);
  }

  _Hypothesis getNew(_Node node, double penaltyToAdd, int index) {
    return _Hypothesis(node, distance + penaltyToAdd, index);
  }

  bool get isWord => node.endsWord;

  @override
  int get hashCode {
    var result = node.hashCode;
    result = result * 31 + distance.hashCode;
    result = result * 31 + index;
    return result;
  }

  @override
  bool operator ==(other) {
    if (other is! _Hypothesis) return false;
    return index == other.index &&
        distance.compareTo(other.distance) == 0 &&
        node == other.node;
  }
}

/// Result value
class Result implements Comparable<Result> {
  final String word;
  final double distance;

  const Result(this.word, this.distance);

  @override
  int compareTo(Result other) => distance.compareTo(other.distance);

  @override
  String toString() => '$word:$distance';
}


lib/dart-spell.dart
import 'dart:collection';

int _nodeIndexCounter=0;

class SingleWordSpellChecker {
  bool checkNearKeySubstitution=false;
  
  final double INSERTION_PENALTY = 1.0;
  final double DELETION_PENALTY = 1.0;
  final double SUBSTITUTION_PENALTY = 1.0;
  final double NEAR_KEY_SUBSTITUTION_PENALTY = 0.5;
  final double TRANSPOSITION_PENALTY = 1.0;
  
  Map<int, String> nearKeyMap = new Map();
  double maxPenalty=0.0;
  Node root;
  
  SingleWordSpellChecker(this.maxPenalty) {
    root = new Node(_nodeIndexCounter++,0);
  }

  void addWord(String word) {    
    //TODO: locale aware lower casing is required.
    addChar(root, 0, word.toLowerCase(), word);
  }

  void addWords(List<String> words) {
    for (String word in words) {
      addWord(word);
    }
  }

  Node addChar(Node currentNode, int index, String word, String actual) {
    int c = word.codeUnitAt(index);
    Node child = currentNode.addChild(c);
    if (index == word.length - 1) {
      child.word = actual;
      return child;
    }
    index++;
    return addChar(child, index, word, actual);
  }
  
  Map<String,double> decode(String input) {
    Hypothesis hyp = new Hypothesis(null, root, 0.0, -1, Hypothesis.N_A);
    Map<String,double> hypotheses = new Map();
    Set<Hypothesis> next = expand(hyp, input, hypotheses);
    while (true) {
      Set<Hypothesis> newHyps = new Set();
      for (Hypothesis hypothesis in next) {
        newHyps.addAll(expand(hypothesis, input, hypotheses));
      }
      if (newHyps.length == 0)
        break;
      next = newHyps;
    }
    return hypotheses;
  }  
  
  Set<Hypothesis> expand(Hypothesis hypothesis, String input, Map<String,double> finished) {
  
    Set<Hypothesis> newHypotheses = new Set();
  
    int nextIndex = hypothesis.index + 1;
  
    // no-error
    if (nextIndex < input.length) {
      if (hypothesis.node.hasChild(input.codeUnitAt(nextIndex))) {
        Hypothesis hyp = hypothesis.getNewMoveForward(
            hypothesis.node.getChild(input.codeUnitAt(nextIndex)),
            0.0,
            Hypothesis.NE);
        if (nextIndex >= input.length - 1) {
          if (hyp.node.word != null)
            addHypothesis(finished, hyp);
        } // TODO: below line may produce unnecessary hypotheses.
        newHypotheses.add(hyp);
      }
    } else if (hypothesis.node.word != null)
      addHypothesis(finished, hypothesis);
  
    // we don't need to explore further if we reached to max penalty
    if (hypothesis.penalty >= maxPenalty)
      return newHypotheses;
  
    // substitution
    if (nextIndex < input.length) {
      for (Node childNode in hypothesis.node.getChildNodes()) {
  
        double penalty = 0.0;
        if (checkNearKeySubstitution) {
          int nextChar = input.codeUnitAt(nextIndex);
          if (childNode.chr != nextChar) {
            String nearCharactersString = nearKeyMap[childNode.chr];
            if (nearCharactersString != null && containsCodeunit(nearCharactersString,nextChar))
              penalty = NEAR_KEY_SUBSTITUTION_PENALTY;
            else penalty = SUBSTITUTION_PENALTY;
          }
        } else penalty = SUBSTITUTION_PENALTY;
  
        if (penalty > 0 && hypothesis.penalty + penalty <= maxPenalty) {
          Hypothesis hyp = hypothesis.getNewMoveForward(
              childNode,
              penalty,
              Hypothesis.SUB);
          if (nextIndex == input.length - 1) {
            if (hyp.node.word != null)
              addHypothesis(finished, hyp);
          } else
            newHypotheses.add(hyp);
        }
      }
    }
  
    if (hypothesis.penalty + DELETION_PENALTY > maxPenalty)
      return newHypotheses;
  
    // deletion
    newHypotheses.add(hypothesis.getNewMoveForward(hypothesis.node, DELETION_PENALTY, Hypothesis.DEL));
  
    // insertion
    for (Node childNode in hypothesis.node.getChildNodes()) {
      newHypotheses.add(hypothesis.getNew(childNode, INSERTION_PENALTY, hypothesis.index, Hypothesis.INS));
    }
  
    // transposition
    if (nextIndex < input.length - 1) {
      int transpose = input.codeUnitAt(nextIndex + 1);
      Node nextNode = hypothesis.node.getChild(transpose);
      int nextChar = input.codeUnitAt(nextIndex);
      if (hypothesis.node.hasChild(transpose) && nextNode.hasChild(nextChar)) {
        Hypothesis hyp = hypothesis.getNew(
            nextNode.getChild(nextChar),
            TRANSPOSITION_PENALTY,
            nextIndex + 1,
            Hypothesis.TR);
        if (nextIndex == input.length - 1) {
          if (hyp.node.word != null)
            addHypothesis(finished, hyp);
        } else
          newHypotheses.add(hyp);
      }
    }
    return newHypotheses;
  }
  
  bool containsCodeunit(String s, int i) {
    for(int c in s.codeUnits) {
      if(c==i)
        return true;
    }
    return false;
  }
  
  addHypothesis(Map<String,double> result, Hypothesis hypothesis) {
    String hypWord = hypothesis.node.word;
    if (hypWord == null) {
      return;
    }
    if (!result.containsKey(hypWord)) {
      result[hypWord]=hypothesis.penalty;
    } else if (result[hypWord] > hypothesis.penalty) {
      result[hypWord]=hypothesis.penalty;
    }
  }
}

class Node {
  int index;
  int chr;
  Map<int, Node> nodes = new Map();
  String word;
  
  Node(this.index, this.chr);
  
  Iterable<Node> getChildNodes() {
    return nodes.values;
  }

  bool hasChild(int c) {
    return nodes.containsKey(c);
  }

  Node getChild(int c) {
    return nodes[c];
  }  
  
  Node addChild(int c) {
    Node node = nodes[c];
    if (node == null) {
      node = new Node(_nodeIndexCounter++, c);
      nodes[c]=node;      
    }
    return node;
  }  
  
  bool operator ==(other) {
    if (!(other is Node)) return false;
    return (index == other.index);
  }  
  
  int get hashCode => index; 
  
}

class Hypothesis implements Comparable<Hypothesis> {
  static const NE = 0;
  static const INS = 1;
  static const DEL = 2;  
  static const SUB = 3;  
  static const TR = 4;  
  static const N_A= 5;
  
  int operation = N_A;
  Hypothesis previous;
  Node node;
  double penalty;
  int index = -1;  
     
  Hypothesis(this.previous, this.node, this.penalty, this.index, this.operation);  
  
  int compareTo(Hypothesis other) => penalty.compareTo(other.penalty);
  
  Hypothesis getNewMoveForward(Node node, double penaltyToAdd, int operation) {
    return new Hypothesis(this, node, this.penalty + penaltyToAdd, index + 1, operation);    
  }  
  
  Hypothesis getNew(Node node, double penaltyToAdd, int index, int operation) {
    return new Hypothesis(this, node, this.penalty + penaltyToAdd, index, operation);    
  }    
  
  bool operator ==(other) {
    if (other is! Hypothesis) return false;
    return index ==  other.index && 
        penalty.compareTo(other.penalty)==0 && 
        node==other.node;    
  }    
  
  int get hashCode {
    int result = node.hashCode;
    result= result*31+penalty.hashCode;
    result= result*31+index;
    return result;
  }    
}

main() {
  var spellChecker = new SingleWordSpellChecker(1.0);
  spellChecker.addWords(["apple","apples","appl"]);
  Map<String,double> hypotheses = spellChecker.decode("apples");
  print (hypotheses);
}